raw_tiles/gen.py

```diff
@@ -19,9 +19,14 @@ def __call__(self, tile):
             buf = StringIO()
             writer = self.formatter.create(buf)
             for record in source_location.records:
-                writer.write(record)
-            writer.close()
+                # record is a tuple here, and writer.write takes varargs, so
+                # we need to unpack the tuple.
+                writer.write(*record)
+            writer.flush()
             data = buf.getvalue()
+            # close writer after we get the buffer value, so that the writer
+            # hasn't closed the buffer yet.
```

Comment:
Is the timing important here? We know that it's an in memory buffer, so as long as everything is flushed first does it matter if we close before vs after calling `getvalue()`?

Reply:
It seems that the previous code worked because the `Gzip` formatter swallowed the call to `close` without closing the underlying file object, but did flush. When using `Msgpack` without `Gzip`, the call to `buf.getvalue()` would fail because the `buf` object had been closed.

Since the structure of the chain is hidden (it's constructed by the formatter objects), I think it's necessary to pass the call to `close` along. Hopefully with the two-stage `flush` then `close`, we don't miss out on any data (the `Gzip` does internal buffering, so we must call `flush`), but ensure that the files are actually closed.

Reply:
Ah, you're right. I didn't realize that calling `.getvlaue()` on a `StringIO` object that had been closed fails.